proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1203 @@
+      i64 => i64(bits),
+      f32 => f32(f32::from_bits(bits as i32)),
+      f64 => f64(f64::from_bits(bits)),
+    })
+  return results
+
+def write_wasm_results(dst, ptr, vals):
+  assert_or_trap(ptr % 8 == 0)
+  for i, val in vals:
+    bits = match val
+      i32(v) => i64::from(v),
+      i64(v) => v,
+      f32(v) => i64::from(v.to_bits())
+      f64(v) => i64::from(v.to_bits())
+    write_8_bytes(dst, ptr + i * 8, bits)
+```
+
+Here `fuse` performs the necessary lifting and lowering to translate all
+parameters and results across the call. This `fuse` function is expected to be
+implemented by "trusted code" since it refers to `TABLES`.
+
+It's worth pointing out that either `instance_a` or `instance_b` could in theory
+be a host with this function as well. A host is more privileged than an
+arbitrary wasm module, so much of this can be specialized when dealing with
+hosts. For example a host might not actually `lower` the parameters and instead
+just work with the raw `lift`-ed values. This means that a host doesn't actually
+need to copy a `string` into host memory, it can reference the raw memory
+sitting in wasm (assuming no multithreading of course). Similarly if a host
+calls an exported function on a wasm module it would likely skip `lift`-ing
+parameters and it would go straight to `lower`-ing each one.
+
+It's also worth pointing out that the order of operations here is intended to be
+canonical. Lifting and lowering can be side-effectful because of calls to
+`canonical_abi_realloc` and `canonical_abi_free`, and the canonical ABI intends
+for the ordering of these calls to indeed be canonical. An important property,
+however, is that lifting import parameters is guaranteed to be side-effect-free
+meaning that host implementations of imported functions can retain all their
+raw-in-memory-pointers as necessary.
+
+# Intrinsics
+
+One of the final details of the canonical ABI is that there will be a few
+intrinsic functions to operate on the various types. These are primarily
+intended for handle-related types and lists.
+
+## Memory Intrinsics
+
+When lowering a list, the canonical ABI needs to allocate space in the
+destination core module's linear memory that can be written into. When lifting
+a returned list, the canonical ABI needs to free the allocated memory after it
+has been read from. When either of these are required, the core module (that's
+calling the canonical ABI) must export one or both of the following functions:
+
+```wasm
+(module
+  (func (export "canonical_abi_realloc") (param i32 i32 i32 i32) (result i32))
+  (func (export "canonical_abi_free") (param i32 i32 i32))
+)
+```
+
+The `canonical_abi_realloc` is provided for use today and to future-proof the
+canonical ABI for a future where the canonical ABI is not the only ABI (e.g.
+full expressive adapter functions). Realloc takes three arguments: the original
+pointer, the original size, the original alignment, and the new desired size. It
```

Comment:
Why is the original alignment passed in, when the new alignment for the new allocation is not passed in? It's not clear how the realloc implementation would know the actual alignment.